conjure-java-jackson-serialization/src/test/java/com/palantir/conjure/java/serialization/BinaryBytesDeserializationModuleTest.java

```diff
@@ -0,0 +1,124 @@
+        RecordContainingObject obj = new RecordContainingObject(
+                new byte[] {1, 2, 3}, new byte[] {4, 5, 6}, Bytes.from(new byte[] {7, 8, 9}));
+        byte[] serialized = smileMapper.writeValueAsBytes(obj);
+
+        RecordContainingObject deserialized = smileMapper.readValue(serialized, RecordContainingObject.class);
+        assertThat(deserialized.obj()).isInstanceOf(Bytes.class).isEqualTo(Bytes.from(new byte[] {1, 2, 3}));
+        assertThat(deserialized.byteArray()).containsExactly(obj.byteArray());
+        assertThat(deserialized.bytes()).isEqualTo(obj.bytes());
+    }
+
+    @Test
+    void deserializedBinaryUuidsAreEqual() throws IOException {
+        // UUIDs are serialized as binary by default, and are the case where we have observed equality issues without
+        // this module
+        Set<UUID> uuids = Set.of(new UUID(0, 1), new UUID(2, 3), new UUID(4, 5));
+        byte[] serialized = smileMapper.writeValueAsBytes(uuids);
+
+        Object first = smileMapper.readValue(serialized, Object.class);
+        Object second = smileMapper.readValue(serialized, Object.class);
+        assertThat(first).isEqualTo(second);
+    }
+
+    @Test
+    void deserializesToByteArrayWhenSpecificallyRequested() throws IOException {
+        byte[] value = {1, 2, 3, 4};
+        byte[] serialized = smileMapper.writeValueAsBytes(value);
+
+        byte[] deserialized = smileMapper.readValue(serialized, byte[].class);
+        assertThat(deserialized).isInstanceOf(byte[].class).containsExactly(value);
+    }
+
+    @Test
+    void deserializesCborBinaryAsBytes() throws IOException {
+        ObjectMapper cborMapper = ObjectMappers.newCborClientObjectMapper();
+        cborMapper.registerModule(new BinaryBytesDeserializationModule());
+
+        byte[] value = {1, 2, 3, 4};
+        byte[] serialized = cborMapper.writeValueAsBytes(value);
+
+        Object deserialized = cborMapper.readValue(serialized, Object.class);
+        assertThat(deserialized).isInstanceOf(Bytes.class).isEqualTo(Bytes.from(value));
+    }
+
+    @Test
+    void deserializesYamlBinaryAsBytes() throws IOException {
+        ObjectMapper yamlMapper = YAMLMapper.builder().build();
+        yamlMapper.registerModule(new BinaryBytesDeserializationModule());
+
+        byte[] value = new byte[] {1, 2, 3};
+        String base64Value = new String(Base64.getEncoder().encode(value), StandardCharsets.UTF_8);
+        String yaml = "testData: !!binary " + base64Value;
+
+        Object deserialized = yamlMapper.readValue(yaml, Object.class);
+        assertThat(deserialized).isInstanceOf(Map.class);
+        Map<?, ?> deserializedMap = (Map<?, ?>) deserialized;
+        assertThat(deserializedMap.get("testData")).isInstanceOf(Bytes.class).isEqualTo(Bytes.from(value));
+    }
+
+    @SuppressWarnings({"DangerousRecordArrayField", "ArrayRecordComponent"}) // Intentional
+    private record RecordContainingObject(Object obj, byte[] byteArray, Bytes bytes) {}
```

Comment:
I think we can define this within `deserializesNestedObjectFields` for readability